File: parallel-programming-with-ms-dotnet/CSharp/Appendix A/Examples/Adapter.cs
//===============================================================================
// Microsoft patterns & practices
// Parallel Programming Guide
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// This code released under the terms of the 
// Microsoft patterns & practices license (http://parallelpatterns.codeplex.com/license).
//===============================================================================

using System;
using System.Threading.Tasks;
using Microsoft.Practices.ParallelGuideSamples.Utilities;

namespace Microsoft.Practices.ParallelGuideSamples.RelatedPatterns
{
    public interface IWithFutures
    {
        Task<int> Start();
    }

    public class FuturesBased : IWithFutures
    {
        public Task<int> Start()
        {
            return Task<int>.Factory.StartNew(() =>
                {
                    SampleUtilities.DoCpuIntensiveOperation(2.0);
                    return 42;
                });
        }
    }

    public interface IWithEvents
    {
        void Start();
        event EventHandler<CompletedEventArgs> Completed;
    }
 
    public class EventBased : IWithEvents
    {
        readonly IWithFutures instance = new FuturesBased();

        public void Start()
        {
            Task<int> task = instance.Start();
            task.ContinueWith((t) =>
            {
                var evt = Completed;
                if (evt != null)
                    evt(this, new CompletedEventArgs(t.Result));
            });
        }

        public event EventHandler<CompletedEventArgs> Completed;
    }

    public class CompletedEventArgs : EventArgs
    {
        public int Result { get; private set; }

        public CompletedEventArgs(int result)
        {
            Result = result;
        }
    }
}

File: parallel-programming-with-ms-dotnet/FSharp/Appendix A/RelatedPatterns/Adapter.fs
//===============================================================================
// Microsoft patterns & practices
// Parallel Programming Guide
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// This code released under the terms of the 
// Microsoft patterns & practices license (http://parallelpatterns.codeplex.com/license).
//===============================================================================

namespace Microsoft.Practices.ParallelGuideSamples.RelatedPatterns

open System
open System.Threading.Tasks
open Microsoft.Practices.ParallelGuideSamples.Utilities

// Future based programming model

type IWithFutures =
    abstract Start : unit -> Task<int>

type FuturesBased() =
    interface IWithFutures with
        member x.Start() =
            Task.Factory.StartNew(fun () ->
                SampleUtilities.DoCpuIntensiveOperation(2.0) |> ignore
                42 )

// Event based programming model (adapter for the future based model)

type CompletedEventArgs(result:int) = 
    inherit EventArgs()
    member x.Result = result

type IWithEvents =
    abstract Start : unit -> unit
    abstract Completed : IEvent<CompletedEventArgs>
 
type EventBased() =
    let instance = new FuturesBased() :> IWithFutures
    let completedEvt = new Event<_>()

    interface IWithEvents with
        member x.Completed = completedEvt.Publish
        member x.Start() =
            let task = instance.Start()
            task.ContinueWith(new Action<Task<_>>(fun t ->
                completedEvt.Trigger(new CompletedEventArgs(t.Result)))) |> ignore
